### ISSUE TITLE
Macros for serializing a type as another type

### ISSUE THREAD

**Issue by Boscop** [jonasbb/serde_with#309]
I've been using these macros for a long time, to serialize a type as a different repr type.
They make it easy to serialize types from other crates that aren't `Serialize`/`Deserialize`.
I think it makes sense to include these macros in this crate, what do you think?

```rust
/// serializes a type as a different repr type that it can be converted Into and From
#[macro_export]
macro_rules! serde_as {
	($m:ident, $t:ty, $r:ty) => {
		pub mod $m {
			use super::*;
			use serde::{Deserialize, Deserializer, Serialize, Serializer};

			pub fn serialize<S: Serializer>(x: &$t, serializer: S) -> Result<S::Ok, S::Error> {
				let y: $r = (*x).into();
				y.serialize(serializer)
			}

			pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<$t, D::Error> {
				let y: $r = Deserialize::deserialize(deserializer)?;
				Ok(y.into())
			}
		}
	};
}

/// serializes a type as a different repr type using the given conversion functions
#[macro_export]
macro_rules! serde_conv {
	($m:ident, $t:ty, $ser:expr, $de:expr) => {
		pub mod $m {
			use super::*;
			use serde::{Deserialize, Deserializer, Serialize, Serializer};

			pub fn serialize<S: Serializer>(x: &$t, serializer: S) -> Result<S::Ok, S::Error> {
				let y = $ser(*x);
				y.serialize(serializer)
			}

			pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<$t, D::Error> {
				let y = Deserialize::deserialize(deserializer)?;
				Ok($de(y))
			}
		}
	};
}
```

Example usage:
```rust
serde_conv!(serde_timestamp, DOMTimeStamp, |x: DOMTimeStamp| x.0 as f64, |x: f64| DOMTimeStamp(x as _));
```
And then
```rust
#[serde(with = "serde_timestamp")] 
timestamp: DOMTimeStamp
```

Or:
```rust
serde_conv!(bool_str, bool, |x: &bool| x.to_string(), |x: String| x.parse::<bool>());

serde_conv!(flags_hashmap_bool_str, HashMap<Flag, bool>,
	|x: &HashMap<Flag, bool>| {
		x.iter().map(|(k, v)| (*k, v.to_string()))
			.collect::<HashMap<Flag, String>>()
	},
	|x: HashMap<Flag, String>| {
		x.into_iter().map(|(k, v)| Ok((k, v.parse::<bool>()?)))
			.collect::<Result<HashMap<Flag, bool>>>()
	}
);

serde_conv!(my_enum_i32, MyEnum, |x: &MyEnum| *x as i32, |x: i32| Ok(match x {
	0 => MyEnum::A,
	1 => MyEnum::B,
	2 => MyEnum::C,
	3 => MyEnum::D,
	_ => bail!("MyEnum: {}", x),
}));
```

**Comment by jonasbb** [comment]
Hi, thank you for the idea. They definitely seem in scope for this crate. Both macros create a new module with a `serialize` and a `deserialize` function, such that the module can be used for `#[serde(with = ...]`.

---

Your `serde_as` seems a bit similar to what I scratched in this PR (https://github.com/jonasbb/serde_with/pull/198/files#diff-7579165df73e6abbb874a640fe6974fb89185b3142a565f1c93a7594777275e1R1114-R1142)
```rust
#[serde_as]
#[derive(Debug, PartialEq, Deserialize)]
struct S(#[serde_as(deserialize_as = "FromInto<u32>")] FromDeserializable);
```

This will use `u32::deserialize` to get a value and then call `.into()` to convert it into a `FromDeserializable`.
What do you think of that approach?
The implementation is generic, so the only thing required is applying the attribute in the necessary places. It also integrates with the `#[serde_with::serde_as]` macro such that it also works in containers, e.g., `#[serde_as(deserialize_as = "Vec<FromInto<u32>>")] Vec<FromDeserializable>` should work.

---

Regarding `serde_conv` that looks useful to reduce the boilerplate, especially since the function signatures are a bit complicated.
Is that the exact definition you are using though? It seems there are some type errors. For example this deref `let y = $ser(*x);` and this `Ok($de(y))` creates a `Result<Result<...` since the converter functions already return `Result`.

The serialization uses an infallible lambda but the deserialization uses a fallible one. Is that on purpose?

Maybe it makes sense for the macro to create a type which can be used not only for serde's with but also for `#[serde_with::serde_as]`. I think it would be enough to take a type name instead of a module. The type would have two inherent methods (`serialize`+`deserialize`) such that it works for serde's with and two trait implementations (`SerializeAs`+`DeserializeAs`).

**Comment by Boscop** [comment]
> This will use u32::deserialize to get a value and then call .into() to convert it into a FromDeserializable.
> What do you think of that approach?

Yeah, your approach is better because you only have to annotate the field, and not have to create a module first by calling a macro at a different code location, right?
The fact that it works in containers is also great!
Btw, this reminds me of another issue I had: Large nested arrays: https://github.com/est31/serde-big-array/issues/7
Does this crate support that case? (I haven't used this crate yet.)

> Is that the exact definition you are using though?

Ah sorry, I had 2 different versions of this macro in 2 different crates. (Which is why I even had the idea to suggest adding the macros to this crate, to avoid copy pasting these macros into every project.)

In that case I was using this definition:
```rust
/// serializes a type as a different repr type using the given conversion functions
#[macro_export]
macro_rules! serde_conv {
	($m:ident, $t:ty, $ser:expr, $de:expr) => {
		pub mod $m {
			use super::*;
			use serde::{Deserialize, Deserializer, Serialize, Serializer};

			pub fn serialize<S: Serializer>(x: &$t, serializer: S) -> Result<S::Ok, S::Error> {
				let y = $ser(x);
				y.serialize(serializer)
			}

			pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<$t, D::Error> {
				let y = Deserialize::deserialize(deserializer)?;
				Ok($de(y).map_err(serde::de::Error::custom)?)
			}
		}
	};
}
```

> The serialization uses an infallible lambda but the deserialization uses a fallible one. Is that on purpose?

Yes, whenever I used this macro, only deserialization had the potential of failing.

> Maybe it makes sense for the macro to create a type which can be used not only for serde's with but also for #[serde_with::serde_as]. I think it would be enough to take a type name instead of a module. The type would have two inherent methods (serialize+deserialize) such that it works for serde's with and two trait implementations (SerializeAs+DeserializeAs).

Yes, feel free to adapt my macro to make it integrate with this crate in the way that makes the most sense :)
The idea of the macro is that in cases where there is no `Into`&`From` impl for thirdparty types (and you don't want to write newtypes), you can just provide some short conversion functions. (It's very useful for quick prototype coding.)

**Comment by jonasbb** [comment]
> > This will use u32::deserialize to get a value and then call .into() to convert it into a FromDeserializable.
> > What do you think of that approach?
>
> Yeah, your approach is better because you only have to annotate the field, and not have to create a module first by calling a macro at a different code location, right?

No, you don't need to create a module, the field annotation is enough.

> Btw, this reminds me of another issue I had: Large nested arrays: [est31/serde-big-array#7](https://github.com/est31/serde-big-array/issues/7)
> Does this crate support that case? (I haven't used this crate yet.)

Yes, you can do that. If you have const-generics (1.51+) you can do this. The `_` means to use the types `Serialize`/`Deserialize` implementations. Instead, a converter type, like the above-mentioned `FromInto` could be specified. I just didn't want to comment on that issue only to promote this crate.

```rust
#[serde_as(as = "[[_; 64]; 33]")]
value: [[u8; 64]; 33],
```

> > The serialization uses an infallible lambda but the deserialization uses a fallible one. Is that on purpose?
> 
> Yes, whenever I used this macro, only deserialization had the potential of failing.

Errors during serialization are probably less common anyway.

> Yes, feel free to adapt my macro to make it integrate with this crate in the way that makes the most sense :)
> The idea of the macro is that in cases where there is no `Into`&`From` impl for thirdparty types (and you don't want to write newtypes), you can just provide some short conversion functions. (It's very useful for quick prototype coding.)

Thanks, I will do that. Since you have interest in #198, I try to get that finished too.

> Btw, in my first post, in the code example, you can see that I had to create the `flags_hashmap_bool_str` module in addition to the `bool_str` module. Both just convert between bools and strings, but it was necessary because the macro doesn't support nesting.
> In other words, I couldn't have written:
> 
> ```rust
> struct S {
>     flags: HashMap<Flag, #[serde(with = "bool_str")] bool>,
> }
> ```
> 
> Do you know a way to define the `serde_conv` macro so that it can be used in such a nested way? :)

Yes, you can do this with `#[serde_with::serde_as]`. It solves this specific problem of applying with-attributes inside containers. Also check the documentation if you haven't used the crate yet.
https://docs.rs/serde_with/1.8.1/serde_with/guide/serde_as/index.html

```rust
#[serde_as]
#[derive(Serialize, Deserialize)]
struct S {
	#[serde_as(as = "HashMap<_, DisplayFromStr>")]
    flags: HashMap<Flag, bool>,
}
```

**Comment by jonasbb** [comment]
I just released a new version 1.9.0 with the `FromInto` / `TryFromInto` types and the `serde_conv!` macro. Let me know if there are any problem with the implementations.
